# 비동기 처리

javaScript는 동기식 언어입니다. 한번에 하나의 작업을 수행할 수 있으며 한 작업이 실행되는 동안에는 다른 작업들은 멈추고 기다려야 합니다. 이런 동작을 단일 스레드(싱글 스레드), 동기(Synchronous) 라고 부릅니다.

javaScript에서 비동기식으로 처리되는 작업들은 아래와 같습니다.

- Ajax Web API 요청
- 파일 읽기
- 암호화/복호화
- 작업 예약

그리고 비동기 처리를 하기 위한 방법은 아래 3가지와 같습니다.

- callback
- Promise
- async/await

## callback

콜백 함수로 처리를 하게 된다면 비동기 작업이 많아질 경우 코드가 쉽게 난잡해지게 되었습니다. 아래의 코드와 같이 비동기적으로 처리해야 하는 일이 많아질수록, 코드의 깊이가 계속 깊어지는 현상이 있습니다. 이러한 현상을 Callback Hell (콜백지옥)이라고 부릅니다.

```javascript
function increaseAndPrint(n, callback) {
  setTimeout(() => {
    const increased = n + 1;
    console.log(increased);
    if (callback) {
      callback(increased);
    }
  }, 1000);
}

increaseAndPrint(0, (n) => {
  increaseAndPrint(n, (n) => {
    increaseAndPrint(n, (n) => {
      increaseAndPrint(n, (n) => {
        increaseAndPrint(n, (n) => {
          console.log("끝!");
        });
      });
    });
  });
});
```

## Promise

프로미스는 비동기 작업을 조금 더 편하게 처리 할 수 있도록 ES6 에 도입된 기능입니다.

### Promise객체

```javascript
new Promise();

// Promise 객체안에 익명 함수 생성
// 파라미터 (resolve, reject)
// 각각 리턴값은 Promise.then, Promise.catch 객체
new Promise((resolve, reject) => {
  //성공할때에는 resolve호출
  //Promise.then 객체를 반환
  reslove("성공");
  //실패했을때는 reject 호출
  //Promise.catch 객체를 반환
  reject("성공");
});
```

```javascript
function increaseAndPrint(n) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const value = n + 1;
      if (value === 5) {
        const error = new Error();
        error.name = "ValueIsFiveError";
        reject(error);
        return;
      }
      console.log(value);
      resolve(value);
    }, 1000);
  });
}

increaseAndPrint(0)
  .then(increaseAndPrint)
  .then(increaseAndPrint)
  .then(increaseAndPrint)
  .then(increaseAndPrint)
  .then(increaseAndPrint)
  .catch((e) => {
    console.error(e);
  });
//   .then(n => {
//     return increaseAndPrint(n);
//   })
//   .then(n => {
//     return increaseAndPrint(n);
//   })
//   .then(n => {
//     return increaseAndPrint(n);
//   })
//   .then(n => {
//     return increaseAndPrint(n);
//   })
//   .then(n => {
//     return increaseAndPrint(n);
//   })
//   .catch(e => {
//     console.error(e);
//   });
```

그렇지만 Promise는 아래의 단점이 존재합니다.

- 에러를 잡을 때 몇번째에서 발생했는지 알아내기도 어려움
- 특정 조건에 따라 분기를 나누는 작업도 어렵다
- 특정 값을 공유해가면서 작업을 처리하기도 까다롭다.

## async/await

async/await 문법은 ES8에 해당하는 문법입니다. Promise 를 더욱 쉽게 사용 할 수 있게 해줍니다.
\*Promise 객체를 사용

### 1단계 async 의 반환값

함수앞에 async를 붙여줍니다. 그렇게 되면 반환값은 Promise 객체가 됩니다. 그 후에 async는 성공시 reslove를 대신하여 return으로 처리해줍니다. retrun시 반환값은 동일하게 Promise.then객체이므로 then을 사용하여 비동기 처리를 할 수 있습니다.

```javascript
async function myAsync() {
  return "async";
}

console.log(myAsync()); //Promise{}

myAsync().then((result) => {
  console.log(result); //async
});
```

### 2단계 async함수안에서 Promise await 사용하기

```javascript
async function myAsync() {
  await delayP(3);
  const result = await delayP(3); //I am Promise
  const result = await delayP().then((resl) => {
    return "I am Promise2";
  }); //I am Promise2

  return "async";
}

function delayP(sec) {
  return new Promise((reslove, reject) => {
    setTimeout(() => {
      reslove("I am Promise");
    }, 1000 * sec);
  });
}
```

### 3단계 async함수안에서 일반 함수 await 사용하기

async함수안에서 Promise 말고도 일반함수도 await 쓸수있다.

```javascript
async function myAsync() {
  const result = await delayP(3). //I am Promise
  return "async";
}

function delayP(sec) {
  setTimeout(() => {
    return "I am Promise";
  }, 1000 * sec);
}
```

### 4단계 예외 처리

1. async의 예외 처리: async가 reslove대신 return을 한것 처럼, 예외 처리할때에는 reject대신 throw를 사용한다. throw를 하면 Promise.catch객체를 반환한다.

```javascript
async function myAsync() {
  throw "error";
}

myAsync().catch((result) => {
  console.log(result);
});
```

2. await 예외 처리: async내부에 await함수를 호출 후 예외 처리를 할 떄에는 try, catch 구문을 작성해야 한다.

```javascript
async function myAsync() {
  try {
    await delayP(5);
  } catch (e) {
    console.log(e);
  }
  throw "error";
}

function delayP(sec) {
  return new Promise((reslove, reject) => {
    setTimeout(() => {
      reject("Error");
    }, 1000 * sec);
  });
}
```
